packages/medusa/src/scripts/oas/build-oas.ts

```diff
@@ -0,0 +1,124 @@
+const debug = (...args) => {
+  if (isVerbose) {
+    console.debug(...args)
+  }
+}
+
+const run = async () => {
+  debug("Generate OAS from codebase.")
+
+  const targetDir = path.resolve(__dirname, "../../../", "dist/oas")
+  fs.mkdirSync(targetDir, { recursive: true })
+
+  for (const apiType of ["store", "admin"]) {
+    debug(`Building OAS for ${apiType} api.`)
+    const oas = await getOASFromCodebase(apiType as ApiType)
+    exportOASToJSON(oas, apiType as ApiType, targetDir)
+  }
+
+  debug("OAS successfully generated.")
+}
+
+const getOASFromCodebase = async (apiType: ApiType) => {
+  debug("Parse JSDoc from path and schema definitions.")
+  const gen = await swaggerInline(
+    [
+      path.resolve(__dirname, "../../", "models"),
+      path.resolve(__dirname, "../../", "api/middlewares"),
+      path.resolve(__dirname, "../../", `api/routes/${apiType}`),
+    ],
+    {
+      base: path.resolve(__dirname, "./", `${apiType}-spec3-base.yaml`),
+      format: ".json",
+    }
+  )
+
+  debug("Get OAS JSON object from OAS JSON string.")
+  const oas = await OpenAPIParser.parse(JSON.parse(gen))
+
+  if (!skipJSONSchema) {
+    debug("Parse Class with class-validator for schema definitions.")
+    const api_ = require("../../api")
+    const models_ = require("../../models")
+    const schemas = validationMetadatasToSchemas(getJSONSchemaOptions())
+
+    const jsdocKeys = Object.keys(oas.components.schemas)
+    const classKeys = Object.keys(schemas)
+
+    debug("Augment OAS with schemas from classes.")
+    oas.components.schemas = Object.assign(oas.components.schemas, schemas)
+
+    if (isVerbose) {
+      // List all schemas and their declaration origin.
+      for (const key of Object.keys(oas.components.schemas)) {
+        debug(
+          `${jsdocKeys.includes(key) ? 1 : 0}, ${
+            classKeys.includes(key) ? 1 : 0
+          }, ${key}`
+        )
+      }
+    }
+  }
+
+  debug("Validate generated OAS.")
+  try {
+    await OpenAPIParser.validate(JSON.parse(JSON.stringify(oas)))
+  } catch (err) {
+    debug("OAS validation failed.")
+    console.warn(err)
+  }
+
+  return oas
+}
+
+const exportOASToJSON = (oas, apiType: ApiType, targetDir: string) => {
+  debug("Exporting OAS to JSON.")
+
+  const json = JSON.stringify(oas, null, 2)
+  fs.writeFile(path.resolve(targetDir, `${apiType}.oas.json`), json, (err) => {
+    if (err) {
+      throw err
+    }
+  })
+}
+
+const getJSONSchemaOptions = (): Partial<IOptions> => ({
+  classTransformerMetadataStorage: defaultMetadataStorage,
+  refPointerPrefix: "#/components/schemas/",
+  additionalConverters: {
+    IsNullable: IsNullableJSONSchemaConverter,
+    IsType: IsTypeJSONSchemaConverter,
+    IsGreaterThan: () => {},
+    IsISO8601Duration: () => {},
+    ExactlyOne: (meta, options) => {
+      return {
+        type: "array",
+        items: {
+          type: typeof meta.constraints[0].toString(),
```

Comment:
suggestion: This will always return `string` is that intended? in that case should we set it to string instead? what was the idea?

Reply:
Good catch, I removed the `.toString()`. The idea is to pick an element from the array in order to infer and declare the array's item type. 
```typescript
@Validate(ExactlyOne, [
  1,
  33,
  89,
  100,
])
@IsArray()
@IsNumber({ each: true })
foobar?: number[]
```
Should produce the following:
```json
"foobar": {
  "items": {
    "type": "number"
  },
  "type": "array",
  "enum": [
    "1",
    "33",
    "89",
    "100"
  ]
},
```

Reply:
Yes I see 👍 careful with things like null, it is of type object 😅